Inappropriate Self type name in "use the fully qualified path for the potential candidates" suggestion

Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ef40cf437f8d092234c378e4c9b22d4e

```rust
struct Thing<X>(X);

trait Method<T> {
    fn method(self) -> T;
}

impl<X> Method<i32> for Thing<X> {
    fn method(self) -> i32 { 0 }
}

impl<X> Method<u32> for Thing<X> {
    fn method(self) -> u32 { 0 }
}

fn main() {
    let thing = Thing(true);
    thing.method();
}
```

The current output is:

```console
error[E0283]: type annotations needed
  --> src/main.rs:17:11
   |
17 |     thing.method();
   |     ------^^^^^^--
   |     |     |
   |     |     cannot infer type for type parameter `T` declared on the trait `Method`
   |     this method call resolves to `T`
   |
note: multiple `impl`s satisfying `Thing<bool>: Method<_>` found
  --> src/main.rs:7:1
   |
7  | impl<X> Method<i32> for Thing<X> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
11 | impl<X> Method<u32> for Thing<X> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: use the fully qualified path for the potential candidates
   |
17 |     <Thing<X> as Method<i32>>::method(thing);
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17 |     <Thing<X> as Method<u32>>::method(thing);
   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Neither of the suggestions compiles, because `X` does not refer to anything inside the body of main.

```console
error[E0412]: cannot find type `X` in this scope
  --> src/main.rs:17:12
   |
17 |     <Thing<X> as Method<i32>>::method(thing);
   |            ^ not found in this scope
   |
help: consider importing this struct
   |
1  | use nalgebra::coordinates::X;
   |
```

Instead, a correct suggestion would use any of the following 5 correct possibilities:

- `<Thing<bool> as Method<i32>>::method(thing);`
- `<Thing<_> as Method<i32>>::method(thing);`
- `<_ as Method<i32>>::method(thing);`
- `<Method<i32>>::method(thing);`
- `Method::<i32>::method(thing);`

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -734,19 +734,22 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 if !impl_candidates.is_empty() && e.span.contains(span)
                     && let Some(expr) = exprs.first()
                     && let ExprKind::Path(hir::QPath::Resolved(_, path)) = expr.kind
-                    && let [path_segment] = path.segments
+                    && let [_] = path.segments
                 {
                     let mut eraser = TypeParamEraser(self.tcx);
                     let candidate_len = impl_candidates.len();
                     let suggestions = impl_candidates.iter().map(|candidate| {
                         let candidate = candidate.super_fold_with(&mut eraser);
-                        format!(
-                            "{}::{}({})",
-                            candidate, segment.ident, path_segment.ident
-                        )
+                        vec![
+                            (expr.span.shrink_to_lo(), format!("{}::{}(", candidate, segment.ident)),
+                            if exprs.len() == 1 {
+                                (expr.span.shrink_to_hi().with_hi(e.span.hi()), ")".to_string())
+                            } else {
+                                (expr.span.shrink_to_hi().with_hi(exprs[1].span.lo()), ", ".to_string())
+                            },
+                        ]
                     });
-                    err.span_suggestions(
-                        e.span,
+                    err.multipart_suggestions(
                         &format!(
                             "use the fully qualified path for the potential candidate{}",
                             pluralize!(candidate_len),

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -738,7 +738,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 {
                     let mut eraser = TypeParamEraser(self.tcx);
                     let candidate_len = impl_candidates.len();
-                    let suggestions = impl_candidates.iter().map(|candidate| {
+                    let mut suggestions: Vec<_> = impl_candidates.iter().map(|candidate| {
                         let candidate = candidate.super_fold_with(&mut eraser);
                         vec![
                             (expr.span.shrink_to_lo(), format!("{}::{}(", candidate, segment.ident)),
@@ -748,13 +748,14 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 (expr.span.shrink_to_hi().with_hi(exprs[1].span.lo()), ", ".to_string())
                             },
                         ]
-                    });
+                    }).collect();
+                    suggestions.sort_by(|a, b| a[0].1.cmp(&b[0].1));
                     err.multipart_suggestions(
                         &format!(
                             "use the fully qualified path for the potential candidate{}",
                             pluralize!(candidate_len),
                         ),
-                        suggestions,
+                        suggestions.into_iter(),
                         Applicability::MaybeIncorrect,
                     );
                 }

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -736,8 +736,10 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     && let ExprKind::Path(hir::QPath::Resolved(_, path)) = expr.kind
                     && let [path_segment] = path.segments
                 {
+                    let mut eraser = TypeParamEraser(self.tcx);
                     let candidate_len = impl_candidates.len();
                     let suggestions = impl_candidates.iter().map(|candidate| {
+                        let candidate = candidate.super_fold_with(&mut eraser);
                         format!(
                             "{}::{}({})",
                             candidate, segment.ident, path_segment.ident
@@ -1033,6 +1035,21 @@ impl<'tcx> TypeFolder<'tcx> for ErrTypeParamEraser<'tcx> {
     fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
         match t.kind() {
             ty::Error(_) => self.tcx().mk_ty_var(ty::TyVid::from_u32(0)),
+            _ => t.super_fold_with(self),
+        }
+    }
+}
+
+/// Replace type parameters with `ty::Infer(ty::Var)` to display `_`.
+struct TypeParamEraser<'tcx>(TyCtxt<'tcx>);
+
+impl<'tcx> TypeFolder<'tcx> for TypeParamEraser<'tcx> {
+    fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+        self.0
+    }
+    fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
+        match t.kind() {
+            ty::Param(_) | ty::Error(_) => self.tcx().mk_ty_var(ty::TyVid::from_u32(0)),
             _ => t.super_fold_with(self),
         }
     }

--- a/src/test/ui/error-codes/E0283.stderr
+++ b/src/test/ui/error-codes/E0283.stderr
@@ -14,7 +14,6 @@ LL |     let bar = foo_impl.into() * 1u32;
    |               |        |
    |               |        cannot infer type for type parameter `T` declared on the trait `Into`
    |               this method call resolves to `T`
-   |               help: use the fully qualified path for the potential candidate: `<Impl as Into<u32>>::into(foo_impl)`
    |
 note: multiple `impl`s satisfying `Impl: Into<_>` found
   --> $DIR/E0283.rs:17:1
@@ -24,6 +23,10 @@ LL | impl Into<u32> for Impl {
    = note: and another `impl` found in the `core` crate:
            - impl<T, U> Into<U> for T
              where U: From<T>;
+help: use the fully qualified path for the potential candidate
+   |
+LL |     let bar = <Impl as Into<u32>>::into(foo_impl) * 1u32;
+   |               ++++++++++++++++++++++++++        ~
 
 error: aborting due to 2 previous errors
 

--- a/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs
+++ b/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs
@@ -1,20 +1,20 @@
 struct Thing<X>(X);
 
 trait Method<T> {
-    fn method(self) -> T;
+    fn method(self, _: i32) -> T;
 }
 
 impl<X> Method<i32> for Thing<X> {
-    fn method(self) -> i32 { 0 }
+    fn method(self, _: i32) -> i32 { 0 }
 }
 
 impl<X> Method<u32> for Thing<X> {
-    fn method(self) -> u32 { 0 }
+    fn method(self, _: i32) -> u32 { 0 }
 }
 
 fn main() {
     let thing = Thing(true);
-    thing.method();
+    thing.method(42);
     //~^ ERROR type annotations needed
     //~| ERROR type annotations needed
 }

--- a/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs
+++ b/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs
@@ -1,0 +1,20 @@
+struct Thing<X>(X);
+
+trait Method<T> {
+    fn method(self) -> T;
+}
+
+impl<X> Method<i32> for Thing<X> {
+    fn method(self) -> i32 { 0 }
+}
+
+impl<X> Method<u32> for Thing<X> {
+    fn method(self) -> u32 { 0 }
+}
+
+fn main() {
+    let thing = Thing(true);
+    thing.method();
+    //~^ ERROR type annotations needed
+    //~| ERROR type annotations needed
+}

--- a/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.stderr
+++ b/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.stderr
@@ -1,0 +1,37 @@
+error[E0282]: type annotations needed
+  --> $DIR/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs:17:11
+   |
+LL |     thing.method();
+   |     ------^^^^^^--
+   |     |     |
+   |     |     cannot infer type for type parameter `T` declared on the trait `Method`
+   |     this method call resolves to `T`
+
+error[E0283]: type annotations needed
+  --> $DIR/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs:17:11
+   |
+LL |     thing.method();
+   |     ------^^^^^^--
+   |     |     |
+   |     |     cannot infer type for type parameter `T` declared on the trait `Method`
+   |     this method call resolves to `T`
+   |
+note: multiple `impl`s satisfying `Thing<bool>: Method<_>` found
+  --> $DIR/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs:7:1
+   |
+LL | impl<X> Method<i32> for Thing<X> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+LL | impl<X> Method<u32> for Thing<X> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+help: use the fully qualified path for the potential candidates
+   |
+LL |     <Thing<_> as Method<i32>>::method(thing);
+   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+LL |     <Thing<_> as Method<u32>>::method(thing);
+   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0282, E0283.
+For more information about an error, try `rustc --explain E0282`.

--- a/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.stderr
+++ b/src/test/ui/traits/do-not-mention-type-params-by-name-in-suggestion-issue-96292.stderr
@@ -1,8 +1,8 @@
 error[E0282]: type annotations needed
   --> $DIR/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs:17:11
    |
-LL |     thing.method();
-   |     ------^^^^^^--
+LL |     thing.method(42);
+   |     ------^^^^^^----
    |     |     |
    |     |     cannot infer type for type parameter `T` declared on the trait `Method`
    |     this method call resolves to `T`
@@ -10,8 +10,8 @@ LL |     thing.method();
 error[E0283]: type annotations needed
   --> $DIR/do-not-mention-type-params-by-name-in-suggestion-issue-96292.rs:17:11
    |
-LL |     thing.method();
-   |     ------^^^^^^--
+LL |     thing.method(42);
+   |     ------^^^^^^----
    |     |     |
    |     |     cannot infer type for type parameter `T` declared on the trait `Method`
    |     this method call resolves to `T`
@@ -26,10 +26,10 @@ LL | impl<X> Method<u32> for Thing<X> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 help: use the fully qualified path for the potential candidates
    |
-LL |     <Thing<_> as Method<i32>>::method(thing);
-   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     <Thing<_> as Method<u32>>::method(thing);
-   |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+LL |     <Thing<_> as Method<i32>>::method(thing, 42);
+   |     ++++++++++++++++++++++++++++++++++     ~
+LL |     <Thing<_> as Method<u32>>::method(thing, 42);
+   |     ++++++++++++++++++++++++++++++++++     ~
 
 error: aborting due to 2 previous errors
 

--- a/src/test/ui/traits/issue-77982.stderr
+++ b/src/test/ui/traits/issue-77982.stderr
@@ -36,14 +36,14 @@ LL |     opts.get(opt.as_ref());
            - impl AsRef<str> for String;
 help: use the fully qualified path for the potential candidates
    |
-LL |     opts.get(<str as AsRef<Path>>::as_ref(opt));
-   |              +++++++++++++++++++++++++++++   ~
-LL |     opts.get(<str as AsRef<OsStr>>::as_ref(opt));
-   |              ++++++++++++++++++++++++++++++   ~
-LL |     opts.get(<str as AsRef<str>>::as_ref(opt));
-   |              ++++++++++++++++++++++++++++   ~
-LL |     opts.get(<str as AsRef<[u8]>>::as_ref(opt));
-   |              +++++++++++++++++++++++++++++   ~
+LL |     opts.get(<String as AsRef<OsStr>>::as_ref(opt));
+   |              +++++++++++++++++++++++++++++++++   ~
+LL |     opts.get(<String as AsRef<Path>>::as_ref(opt));
+   |              ++++++++++++++++++++++++++++++++   ~
+LL |     opts.get(<String as AsRef<[u8]>>::as_ref(opt));
+   |              ++++++++++++++++++++++++++++++++   ~
+LL |     opts.get(<String as AsRef<str>>::as_ref(opt));
+   |              +++++++++++++++++++++++++++++++   ~
      and 4 other candidates
 
 error[E0283]: type annotations needed

--- a/src/test/ui/traits/issue-77982.stderr
+++ b/src/test/ui/traits/issue-77982.stderr
@@ -36,14 +36,14 @@ LL |     opts.get(opt.as_ref());
            - impl AsRef<str> for String;
 help: use the fully qualified path for the potential candidates
    |
-LL |     opts.get(<String as AsRef<OsStr>>::as_ref(opt));
-   |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     opts.get(<String as AsRef<Path>>::as_ref(opt));
-   |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     opts.get(<String as AsRef<[u8]>>::as_ref(opt));
-   |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     opts.get(<String as AsRef<str>>::as_ref(opt));
-   |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+LL |     opts.get(<str as AsRef<Path>>::as_ref(opt));
+   |              +++++++++++++++++++++++++++++   ~
+LL |     opts.get(<str as AsRef<OsStr>>::as_ref(opt));
+   |              ++++++++++++++++++++++++++++++   ~
+LL |     opts.get(<str as AsRef<str>>::as_ref(opt));
+   |              ++++++++++++++++++++++++++++   ~
+LL |     opts.get(<str as AsRef<[u8]>>::as_ref(opt));
+   |              +++++++++++++++++++++++++++++   ~
      and 4 other candidates
 
 error[E0283]: type annotations needed

